src/mpid/ch4/netmod/ofi/ofi_rndv_read.c

```diff
@@ -0,0 +1,476 @@
+        MPIDI_OFI_CALL(fi_close(&p->u.send.mrs[i]->fid), mr_unreg);
+        if (!MPIDI_OFI_ENABLE_MR_PROV_KEY) {
+            MPIDI_OFI_mr_key_free(MPIDI_OFI_LOCAL_MR_KEY, key);
+        }
+    }
+    MPL_free(p->u.send.mrs);
+    MPL_free(r);
+
+    /* complete sreq */
+    MPIR_Datatype_release_if_not_builtin(p->datatype);
+    MPIDI_Request_complete_fast(sreq);
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* -- receiver side -- */
+
+int MPIDI_OFI_rndvread_recv(MPIR_Request * rreq, int tag, int vci_src, int vci_dst)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_OFI_rndvread_t *p = &MPIDI_OFI_AMREQ_READ(rreq);
+    MPIR_FUNC_ENTER;
+
+    p->buf = MPIDIG_REQUEST(rreq, buffer);
+    p->count = MPIDIG_REQUEST(rreq, count);
+    p->datatype = MPIDIG_REQUEST(rreq, datatype);
+    p->vci_local = vci_dst;
+    p->vci_remote = vci_src;
+    p->av = MPIDIU_comm_rank_to_av(rreq->comm, rreq->status.MPI_SOURCE);
+    p->match_bits = MPIDI_OFI_init_sendtag(rreq->comm->recvcontext_id, 0, tag) | MPIDI_OFI_AM_SEND;
+    p->num_nics = MPIDI_OFI_global.num_nics;
+
+    MPIR_GPU_query_pointer_attr(p->buf, &p->attr);
+
+    MPIR_Datatype_get_size_macro(p->datatype, p->data_sz);
+    p->data_sz *= p->count;
+
+    int dt_contig;
+    MPIR_Datatype_is_contig(p->datatype, &dt_contig);
+    p->u.recv.need_pack = need_pack(dt_contig, &p->attr);
+
+    if (!p->u.recv.need_pack) {
+        MPI_Aint true_extent, true_lb;
+        MPIR_Type_get_true_extent_impl(p->datatype, &true_lb, &true_extent);
+        p->u.recv.u.data = MPIR_get_contig_ptr(p->buf, true_lb);
+    } else {
+        p->u.recv.u.copy_infly = 0;
+    }
+
+    /* recv the mrs */
+    MPI_Aint hdr_sz = sizeof(struct rdma_info) + p->num_nics * sizeof(uint64_t);
+    mpi_errno = MPIDI_OFI_RNDV_recv_hdr(rreq, MPIDI_OFI_EVENT_RNDVREAD_RECV_MRS, hdr_sz,
+                                        p->av, p->vci_local, p->vci_remote, p->match_bits);
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPIR_FUNC_EXIT;
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_OFI_rndvread_recv_mrs_event(struct fi_cq_tagged_entry *wc, MPIR_Request * r)
+{
+    int mpi_errno = MPI_SUCCESS;
+    struct rdma_info *hdr = MPIDI_OFI_RNDV_GET_CONTROL_HDR(r);
+    MPIR_Request *rreq = MPIDI_OFI_RNDV_GET_CONTROL_REQ(r);
+    MPIDI_OFI_rndvread_t *p = &MPIDI_OFI_AMREQ_READ(rreq);
+
+    MPIR_Assert(hdr->num_nics == p->num_nics);
+    p->u.recv.remote_data_sz = hdr->data_sz;
+    p->u.recv.remote_base = hdr->base;
+    p->u.recv.rkeys = MPL_malloc(p->num_nics * sizeof(uint64_t), MPL_MEM_OTHER);
+    for (int i = 0; i < p->num_nics; i++) {
+        p->u.recv.rkeys[i] = hdr->rkeys[i];
+    }
+
+    MPL_free(r);
+
+    /* setup chunks */
+    p->u.recv.chunks_per_nic = get_chunks_per_nic(p->u.recv.remote_data_sz, p->num_nics);
+
+    p->u.recv.cur_chunk_index = 0;
+    p->u.recv.num_infly = 0;
+
+    /* issue fi_read */
+    mpi_errno = MPIR_Async_things_add(rndvread_read_poll, rreq, NULL);
```

Comment:
this is just my ignorance of the async things interface, but why do we only "add" items in the rndv read protocol, but we "spawn" them in the pipelined send protocol?

Reply:
If it is creating new async task within an async callback, we need use "spawn". This is to prevent corrupting the task list and avoiding recursive lock.

Reply:
Not really. Are you available for offline chat?

Reply:
Read ops are async tasks but they are ofi async tasks and managed by libfabric. The callback for ofi async tasks are those event functions. Our MPIX Async facility manages "local" async tasks (or all async tasks that requires manual polls).